kv.go

```diff
@@ -0,0 +1,387 @@
+func (kv *kvs) Create(key string, value []byte) (revision uint64, err error) {
+	return kv.Update(key, value, 0)
+}
+
+// Update will update the value iff the latest revision matches.
+func (kv *kvs) Update(key string, value []byte, revision uint64) (uint64, error) {
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	m := Msg{Subject: b.String(), Header: Header{}, Data: value}
+	m.Header.Set(ExpectedLastSubjSeqHdr, strconv.FormatUint(revision, 10))
+
+	pa, err := kv.js.PublishMsg(&m)
+	if err != nil {
+		return 0, err
+	}
+	return pa.Sequence, err
+}
+
+// Delete the key and all revisions.
+func (kv *kvs) Delete(key string) error {
+	o, cancel, err := getJSContextOpts(kv.js.opts)
+	if err != nil {
+		return err
+	}
+	if cancel != nil {
+		defer cancel()
+	}
+
+	// Build by hand since simple and avoids stdlib JSON.
+	var b strings.Builder
+	b.WriteString("{\"filter\":\"")
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+	b.WriteString("\"}")
+
+	// Send request.
+	subj := kv.js.apiSubj(fmt.Sprintf(apiStreamPurgeT, kv.stream))
+	r, err := kv.nc.RequestWithContext(o.ctx, subj, []byte(b.String()))
+	if err != nil {
+		return err
+	}
+	var resp streamPurgeResponse
+	if err := json.Unmarshal(r.Data, &resp); err != nil {
+		return err
+	}
+	if resp.Error != nil {
+		return errors.New(resp.Error.Description)
+	}
+	return nil
+}
+
+// List will return all values for the key.
+func (kv *kvs) List(key string) ([]*KeyValueEntry, error) {
+	o, cancel, err := getJSContextOpts(kv.js.opts)
+	if err != nil {
+		return nil, err
+	}
+	if cancel != nil {
+		defer cancel()
+	}
+
+	ctx := o.ctx
+	if ctx == nil {
+		ctx, cancel = context.WithTimeout(context.Background(), o.wait)
+		defer cancel()
+	}
+
+	var vals []*KeyValueEntry
+	done := make(chan error, 1)
+	cb := func(m *Msg) {
+		tokens, err := getMetadataFields(m.Reply)
+		if err != nil {
+			done <- err
+		} else {
+			if len(m.Subject) <= len(kv.pre) {
+				done <- ErrBadSubject
+				return
+			}
+			subj := m.Subject[len(kv.pre):]
+			vals = append(vals, &KeyValueEntry{
+				Key:      subj,
+				Data:     m.Data,
+				Revision: uint64(parseNum(tokens[ackStreamSeqTokenPos])),
+			})
+			if tokens[ackNumPendingTokenPos] == kvNoPending {
+				done <- nil
+			}
+		}
+	}
+
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	// Used ordered consumer to deliver results.
+	sub, err := kv.js.Subscribe(b.String(), cb, OrderedConsumer())
+	if err != nil {
+		return nil, err
+	}
+	defer sub.Unsubscribe()
+
+	// Wait on done or ctx/timeout.
+	select {
+	case err := <-done:
+		if err != nil {
+			return nil, err
+		}
+		return vals, nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+// WatchAll watches all keys.
+func (kv *kvs) WatchAll(cb KeyValueUpdate) (*Subscription, error) {
+	return kv.Watch(AllKeys, cb)
+}
+
+// Watch will fire the callback when a key that matches the keys pattern is updated.
+// keys needs to be a valid NATS subject.
```

Comment:
but in this implementation - due to not using operation on delete - would be unaware of deletes meaning one could not build a cache or ever actually accurate watched view of a key

Reply:
How would you be aware if I did do deletes?

Reply:
as per ADR - a nil body with the correct header set. I would not know if you did a Purge for example though.

Reply:
Seems this is part of ADR I should go back and take a look at to see if that would actually work.

Reply:
```
$ nats kv put TEST x.y z
z
$ nats kv del TEST x.y -f
$ nats kv get TEST x.y
nats: error: unknown key
$ nats kv history TEST x.y
╭──────────────────────────────────────────────────╮
│              History for TEST > x.y              │
├─────┬─────┬─────────────────────┬────────┬───────┤
│ Seq │ Op  │ Created             │ Length │ Value │
├─────┼─────┼─────────────────────┼────────┼───────┤
│ 1   │ PUT │ 27 Sep 21 14:15 UTC │ 1      │ z     │
│ 2   │ DEL │ 27 Sep 21 14:15 UTC │ 0      │       │
╰─────┴─────┴─────────────────────┴────────┴───────╯
$ ats s get KV_TEST --last-for '$KV.TEST.x.y' --json
{
  "subject": "$KV.TEST.x.y",
  "seq": 2,
  "hdrs": "TkFUUy8xLjANCktWLU9wZXJhdGlvbjogREVMDQoNCg==",
  "time": "2021-09-27T14:15:03.844010066Z"
}